### ISSUE TITLE
Future Features

### ISSUE THREAD

**Issue by luixal** [msavin/SteveJobs...meteor.schedule.background.tasks.jobs.queue#4]
I don't know if this is open to community consideration but... what the hell! I'm going to write down my opinions over the possible new features commented in the readme, just to get more opinnions :)

1. Create a way to run jobs across multiple servers. [...] --> _maybe different collections? so each server can handle its own? Would need to implement kind of a "proxy" in the API but seems better for performance, doesn't it?_
2. Create a way for the server to pause jobs if the CPU usage is high. Every little bit helps, right?  --> _i would implement this after 5, just to allow to run important jobs in any case_
3. Create a way to run jobs as a microservice. --> _maybe that would be another project that uses this package and adds a REST API?_
4. Create a way to repeat jobs. (Perhaps this should be another package?) --> _as there are many formats to represent schedules and repetitions, I would go with different packages, yep_ :+1: 
5. Create a way to prioritize certain jobs --> _nothing to add here :)_
6. Add support for hooks --> _nothing to add here neither :)_
7. Add Jobs.delay() to delaying tasks --> _maybe a field in the job object? a Jobs.delay() to delay all jobs and a Jobs.delay(jobId), Jobs.delay(priorityLevel), ... also related to 5?_
8. Add support for setting manual timezones on jobs --> _using momentJS should be "easy", isn't it?_

bonus track:

9. Add a template you can use as UI in a project's admin are (control job queue, jobs and get execution logs) --> this would totally be another package :)
10. Add some basic functions like: `Jobs.isStarted()` and `Jobs.getStats()` or just a `Jobs.getStatus()` for getting if the queue is stopped/started and jobs pending/failed/succeeded/canceled/running. This would be related to 1 and 5.

I'm totally interested in 4 and quite interested in 7 and 8. Also in 9 and 10 logically. I can put some spare time in this too :)

**Comment by msavin** [comment]
Hey man - love it - thanks for the interest! 

I created new issues to kick off the next couple of features:
1- https://github.com/msavin/SteveJobs-meteor-jobs-queue/issues/6
2- https://github.com/msavin/SteveJobs-meteor-jobs-queue/issues/5

To your points:
7 - Do you mean like, delay all the jobs? Yeah that would be cool, if a job id is not specified it should run on all jobs. 
8- Indeed, though I like to reduce third party dependencies because something almost always goes wrong with them. 
9- yes :D it could be really handy for development purposes too

------

10- This is a tricky one because I am not sure if jobs should have statuses. I'm thinking of jobs in this context as method calls that are being called but are running at a delay. They either return a result or fail and retry. 

However, someone could implement a status system specific their needs. For example, if they are sending 10,000 emails, they can create a Emails collection, insert a document for that action, and then update the document as they need to.  That way, SJ doesn't have expose its MongoDB collection and remains a lean option. And.. this does make me think about before/after hooks. 

```
Jobs.register({
  sendEmails: {
    before: function (sender, content) {
      EmailHistory.insert({ ... })
    },
    on: function (sender, content) {
      Magic.sendEmails({}, function () {
        EmailHistory.magicUpdate({ ... })
      })
    },
    after: function (e, r) {
      if (!e) {
        Magic.sendConfirmationThatEmailsHaveBeenSentSuccessfully();
      }
    },
  }
})
```
In the example above, we assume that you have an application that is sending out a lot of emails on behalf of your end users. When the job is added, `before` runs and adds a document to your Emails collection. When the job runs, it it would update that document with the latest status. Once it completes, it will let you know the error/result and run the appropriate callback.

**Comment by luixal** [comment]
> 7 - Do you mean like, delay all the jobs? Yeah that would be cool, if a job id is not specified it should run on all jobs. It could get performance intensive if the queue has to update 10,000 documents though.

May be delaying all the jobs could be done by delaying just the function that queries for new jobs to run. That's why I was thinking of an specific function for that.

>8- Indeed, though I like to reduce third party dependencies because something almost always goes wrong with them.

Well... maybe using a fixed version of those third party dependencies would be enough. The point is that using third party tools removes a lot of work from the table. i.e: i already have an UI for generating laterjs recurrencies, now, i would need to write some new code for supporting this package instead of using just what I have. Also momentjs makes converting times between timezones just a line of code.

> 9- yes :D it could be really handy for development purposes too

Or just a handy log for jobs running in a production environment :)

> 10- This is a tricky one because I am not sure if jobs should have statuses. I'm thinking of jobs in this context as method calls that are being called but are running at a delay. They either return a result or fail and retry.

Right now you already have two "states": already run and not run yet. I think the _running right now_ would be handy. The before/after hook idea is great and can be used to implement that _running now_ state. I was already thinking about implementing it with the [`collection-hooks`](https://github.com/matb33/meteor-collection-hooks) package.

**Comment by msavin** [comment]
7- hmm yeah that could be trivial to implement 

8- Moment.js may be an exception here, it's a great library and timezones are a crazy thing :D

10- Regarding adding a "running" state - I am still not sold on the idea. The reason is, we add latency to the job, plus I have trust issues with MongoDB. If something goes wrong with the database or the server, the job could get stuck in the "running" state. Handling that adds a lot of complexity compared to the two state method.

**Comment by msavin** [comment]
Although, if we do add some kind of "running" state, we could start running multiple jobs at once. Once the database update is complete, it can start the next job. 

This also paves the way to run jobs on multiple servers, but the problem there would be, preventing jobs from running twice in case one server grabs the job while the second one is claiming it. Maybe there's something that could be done with Mongo commands to prevent this though.

**Comment by luixal** [comment]
> 8- Moment.js may be an exception here, it's a great library and timezones are a crazy thing :D

Totally insane thing. Once I found momentjs, never looked back xD

> This also paves the way to run jobs on multiple servers, but the problem there would be, preventing jobs from running twice in case one server grabs the job while the second one is claiming it. Maybe there's something that could be done with Mongo commands to prevent this though.

It's a important thing. I have had issues with mongo when inserting a bunch of documents consecutively (even if you pass a callback to `insert`, it gets called once the `_id` has been generated, not when the inserting has been totally completed. Not sure if the same thing applies to `update`).

Maybe using a singleton approach for the _job grabber_? That would work when running jobs on a single instance, but would be a problem when running on multiple instances/servers (I've faced those problem when using `cfs` and `job-collection` packages. Found no solution but making them run only on one instance).

Another solutions would be to set a `jobRunnerInstanceId` when adding the job and then make each instance query for it's own instanceId but... won't leverage CPU loads.
